### PR TITLE
C#: Support exporting for Android

### PR DESCRIPTION
**Pull request by RedworkDE**
Limitations: While https://github.com/dotnet/core/blob/main/release-notes/6.0/supported-os.md and https://github.com/dotnet/core/blob/main/release-notes/7.0/supported-os.md claim support for android on x64, arm32 and arm64 for both net6 and net7, actually trying to publish a project for android with the target framework net6 always fails with an error about the native runtime package being missing, same for arm32 on net7. So only x64 and arm64 with target framework net7 can actually be exported.

Because android is particular about what you can put into shared objects, put the published files with the regular assets instead and then copy they on startup into some temporary location so that they can be loaded normally.

Screenshots of a test app running on emulator and an actual tablet:
Screenshot_20230302_163047
Screenshot_20230302-160748_Net7

TODOs:
- [ ] What is the best place to put the published outputs for execution? (Also keeping other platforms in mind, even tho this PR doesn't allow other platforms to use this option yet)
- [x] Don't copy everything on every start, just on the first and when things have changed.
- [x] ~The assets can sometimes (always?) be accessed by some path directly, so don't copy anything if this is possible.~
  This doesn't really make sense / work. I still left one try with `globalize_path` as the cost is pretty negligible, but could save a lot of time for the first startup

### CODE REVIEW

**Comment by ThomasUijlen** [comment]
Does this already work? I'm looking to test C# on Android ASAP :O

**Comment by RedworkDE** [comment]
> Does this already work? I'm looking to test C# on Android ASAP :O

Yes, this does already work in its current state.

> If so, are there any guides on how to compile everything?

[Compiling the editor with mono](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_with_mono.html#generate-the-glue) (note that all the bits that talk about mono itself are outdated and no longer relevant) and [compiling the android template](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_android.html#building-the-export-templates), but add the `module_mono_enabled=yes` argument to the scons commands.

**Comment by MadFlyFish** [comment]
Great job. Thank you.
I really hope it will release with 4.0.

**Comment by Calinou** [comment]
> Great job. Thank you. I really hope it will release with 4.0.

This is targeting a future 4.x release, as 4.0 is currently in release candidate stage.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs [hunk 1]
```diff
@@ -193,7 +228,7 @@ private string DetermineRuntimeIdentifierArch(string arch)
                 "x86_64" => "x64",
                 "armeabi-v7a" => "arm",
                 "arm64-v8a" => "arm64",
-                "armv7" => "arm",
+                "arm32" => "arm",
```
Comment (raulsntos) [inline modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:231]: This seems like an unrelated change that could've been a separate PR. I'm guessing we just missed it because we weren't supporting this architecture before. Seems to be a result of https://github.com/godotengine/godot/pull/55778.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs [hunk 2]
```diff
@@ -169,17 +177,44 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
...
 
+                // Add to the exported project shared object list or packed resources.
                 foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
                 {
-                    AddSharedObject(file, tags: null,
-                        Path.Join(projectDataDirName,
-                            Path.GetRelativePath(publishOutputTempDir, Path.GetDirectoryName(file))));
+                    if (embedBuildResults)
+                    {
+                        var filePath = SanitizeSlashes(Path.GetRelativePath(publishOutputTempDir, file));
```
Comment (raulsntos) [inline modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:187]: I'm curious why sanitizing the path is needed for android exports.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs [hunk 3]
```diff
@@ -108,15 +110,19 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
             {
                 archs.Add("x86_64");
             }
-            else if (features.Contains("x86_32"))
+            if (features.Contains("x86_32"))
```
Comment (raulsntos) [inline modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:113]: Why remove the `else`? Is android similar to mac universal exports?

File: modules/mono/godotsharp_dirs.cpp [hunk 4]
```diff
@@ -139,19 +135,52 @@ class _GodotSharpDirs {
 		String arch = Engine::get_singleton()->get_architecture_name();
 		String appname = GLOBAL_GET("application/config/name");
 		String appname_safe = OS::get_singleton()->get_safe_dir_name(appname);
-		String data_dir_root = exe_dir.path_join("data_" + appname_safe + "_" + arch);
-		if (!DirAccess::exists(data_dir_root)) {
-			data_dir_root = exe_dir.path_join("data_Godot_" + arch);
-		}
+		String packed_path = "res://.godot/mono/publish/" + arch;
+		if (DirAccess::exists(packed_path)) {
+			// The dotnet publish data is packed in the pck/zip.
```
Comment (raulsntos) [inline modules/mono/godotsharp_dirs.cpp:140]: Is this only relevant for Android?

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 3 above)
Comment (RedworkDE) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:113]: Yes, apks can target multiple architectures as well, and thus may need to contain the build for all. Unlike `universal` there is no dedicated feature tag, the export just has multiple tags.

(This reminds me that I wanted to check how exactly `universal` works and if the universal case or the distinct can be removed)

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 2 above)
Comment (RedworkDE) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:187]: Not specific to android, but if the pck/zip contains files using (some) backslashes as the path separator, these files cannot be retrieved, as during lookup it will hash the path using forward slashes, and then be different to the hash with backslashes.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 1 above)
Comment (RedworkDE) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:231]: Could be a different PR, yes. but it doesn't appear to be related to that PR specifically.

The keys are what ends up in the data directly name and must match https://github.com/godotengine/godot/blob/master/core/config/engine.cpp#L196-L230 to load correctly.

File: modules/mono/godotsharp_dirs.cpp (hunk 4 above)
Comment (RedworkDE) [reply modules/mono/godotsharp_dirs.cpp:140]: As this PR is atm yes, but mostly because I didn't want to take a dependency on the export options PR. With that putting the data files into the pck would be relevant for all platforms.

**Comment by neikeq** [comment]
From what I understand, this looks somewhat similar to how it works in 3.x (although much simpler ofc), where we don't target the android specific target framework, right? What .NET runtime is it using, is it Mono?

I don't think this is the right way to go. My plan for Godot 4 was to not do anything too custom that's not well-supported by .NET. I'm not convinced the supported-os page is referring to this when they claim Android support. I don't really know for sure, but I think `net6-android` is the only recommended target.
If possible, I would also like NuGet to work well, and many NuGet packages only work well when targeting the Android variant of the framework.

I'm currently waiting to see what's coming in .NET 8, as I think they are working on support for exporting a `net8-android` project as a self-contained library. But I'm not certain, so we'll have to wait and see.

**Comment by neikeq** [comment]
> actually trying to publish a project for android with the target framework net6 always fails with an error about the native runtime package being missing, same for arm32 on net7. So only x64 and arm64 with target framework net7 can actually be exported.

Does it use the CoreCLR runtime? If so, does that still happen when you pass `-p:UseMonoRuntime=true` to `dotnet`.

**Comment by RedworkDE** [comment]
This currently works by simply specifying the `android-arm64` rid for publishing, and then the output then includes a `libhostfxr.so` that takes care of everything like on the other platforms.

As far as I can tell, these exports use coreclr as the runtime and mono as the jit.

I'll have to test the rest for a bit.

**Comment by RedworkDE** [comment]
Before I forget to comment again:
- `UseMonoRuntime=true` this makes it possible to publish for android with all 4 architectures supported by godot, but those exports use the mono runtime (duh) and thus the CoreCLR loading code no longer works. I did not investigate if it is possible to load this without bringing all the old mono support code back.
- Using the `net6.0-android` or `net7.0-android` tfm requires that you install / configure a whole bunch of android development dependencies, but the publish doesn't actually work and just emits the main managed assembly and no dependencies. Not sure what exactly is happening here.

**Comment by vonwell** [comment]
> From what I understand, this looks somewhat similar to how it works in 3.x (although much simpler ofc), where we don't target the android specific target framework, right? What .NET runtime is it using, is it Mono?
> 
> I don't think this is the right way to go. My plan for Godot 4 was to not do anything too custom that's not well-supported by .NET. I'm not convinced the supported-os page is referring to this when they claim Android support. I don't really know for sure, but I think `net6-android` is the only recommended target. If possible, I would also like NuGet to work well, and many NuGet packages only work well when targeting the Android variant of the framework.
> 
> I'm currently waiting to see what's coming in .NET 8, as I think they are working on support for exporting a `net8-android` project as a self-contained library. But I'm not certain, so we'll have to wait and see.

In other words, does the multi-platform export template of C# have to wait until after net8?

**Comment by neikeq** [comment]
While the solution we're going to go with is a different one, currently it still relies on some workarounds this PR implements (copying and extracting the assemblies). So if you can fix the conflicts, I think it's fine to merge this first, then I add the other bits.

**Comment by RedworkDE** [comment]
I rebased #76305 (which contains the interesting bits with the assemblies handling) and then rebased this PR onto that, tho there isn't really much left. Didn't test this as probably  everything you want to merge is in the other PR.

**Comment by magian1127** [comment]
I suggest merging this PR in 4.1.
This can enhance the confidence of many people.

**Comment by AThousandShips** [comment]
This will not be merged for 4.1 it needs extensive testing and we're not only in feature freeze but also in release freeze

**Comment by XorZy** [comment]
> > Does this already work? I'm looking to test C# on Android ASAP :O
> 
> Yes, this does already work in its current state.
> 
> > If so, are there any guides on how to compile everything?
> 
> [Compiling the editor with mono](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_with_mono.html#generate-the-glue) (note that all the bits that talk about mono itself are outdated and no longer relevant) and [compiling the android template](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_android.html#building-the-export-templates), but add the `module_mono_enabled=yes` argument to the scons commands.

Does this still hold true?
I have tried building the android export template with `module_mono_enabled=yes` as indicated but I am met with the error 
`RuntimeError: This module does not currently support building for this platform:`
I suppose there have been changes since your initial comment but is it still possible to get it to work?
I am currently working on a C# game and being able to export to Android would be extremely beneficial.
I would be willing to provide any help required if this PR can give use Android builds without having to wait for .NET8.

**Comment by AThousandShips** [comment]
Please read the above link

**Comment by XorZy** [comment]
> Please read the above link

I have, hence my question.
This PR (supposedly ?) does it it in a different way and was reported to be working a few months ago, so I'm curious if it can still be used, even out of tree, to circumvent the aforementioned problems with .NET8.

**Comment by MitchMakesThings** [comment]
@XorZy I compiled it in the last week or two and things seem to be working fine.
I'm almost certain you've missed a step from the [ Compiling for Android ](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_android.html#setting-up-the-buildsystem) page. Make sure you've got the ANDROID_SDK_ROOT environment variable set.

I also updated my csproj to .NET 7 instead of 6. Not sure if that was required or not.

**Comment by XorZy** [comment]
> @XorZy I compiled it in the last week or two and things seem to be working fine. I'm almost certain you've missed a step from the [ Compiling for Android ](https://docs.godotengine.org/en/latest/contributing/development/compiling/compiling_for_android.html#setting-up-the-buildsystem) page. Make sure you've got the ANDROID_SDK_ROOT environment variable set.
> 
> I also updated my csproj to .NET 7 instead of 6. Not sure if that was required or not.

Thanks for letting me know!
Yeah it's entirely possible I made a mistake, I am not familiar with the Android build process.
I think I have set the SDK path properly since I can compile it without `module_mono_enabled=yes`.
I compiled it using `RedworkDE:net-android-support` (f759cc0), maybe that was my mistake?

**Comment by emmauss** [comment]
Net8 will have nativeAOT support for Linux bionic both x64 and arm64 variants. Full support for shared libraries has only been available in the dotnet runtime repo for about 2 months, so I'm surprised how far you've gotten on net7. Are you publishing as a self contained app, or shared library?
Early attempts to create a simple dotnet(not mono) app for Android, showed that you can create a self contained executable for the bionic target, but not a shared library. The latter requiring nativeAOT.

**Review by neikeq — APPROVED** [review]

**Comment by neikeq** [comment]
Since we're not yet in feature freeze for 4.2, I'm merging this now so more people can start testing, and since I need it for my PR as well.

**Comment by Zamir7** [comment]
I downloaded godot 4.2 dev3 and I'm trying to export to android, but it says that in version 4 the export is not yet available, although the release says that you can already test the export
[https://i.gyazo.com/64f8d238d4c8d38d64070ac3a622c7a0.png](url)

**Comment by XorZy** [comment]
> I downloaded godot 4.2 dev3 and I'm trying to export to android, but it says that in version 4 the export is not yet available, although the release says that you can already test the export (C#: Add initial support for exporting to Android ([GH-73257](https://github.com/godotengine/godot/pull/73257))
> 
> er

There's already a PR (#80521) fixing that issue, hopefully it will be merged soon.

**Comment by Zamir7** [comment]
after export, it gives an error on the phone (Godot 4.2 dev4)
Screenshot_20230902-155708

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 1 above)
Comment (akien-mga) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:263]: Since `arm32` support seems broken currently, might be worth re-assessing whether we're doing everything right with these remappings to see if the problem is in Godot or .NET.

**Comment by Zamir7** [comment]
> @Zamir7 See the "Limitations" paragraph in the first post. .NET 7.0 doesn't properly support arm32 currently.

how about it

**Comment by Zamir7** [comment]
On .Net 7, the export is successful, but the application does not function, although it starts, practically nothing works, no matter what you click on the screen, nothing changes, as if the ui buttons do not work, and not only ui.

File: modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs (hunk 1 above)
Comment (raulsntos) [reply modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs:263]: This remap should result in the runtime ID `android-arm` which is correct AFAIK: https://github.com/dotnet/runtime/blob/25c675ff78e0446fe596cea25c7e3969b0936a33/src/libraries/Microsoft.NETCore.Platforms/src/runtime.json#L238

However, the package `Microsoft.NETCore.App.Runtime.linux-bionic-arm` does not seem to exist, so it seems like it's just unsupported in .NET. The runtime ID `android-arm` is also not listed in the [Known RIDs](https://learn.microsoft.com/en-us/dotnet/core/rid-catalog#known-rids).

**Comment by ozanyasindogan** [comment]
One of the biggest challenge for me using C# with Godot now for Android deployments is that you can't make use any .Net socket function which requires SSL/TLS. The app crashes immediately. It's working fine on Windows though. It means you can't use HttpClient for https, SSLStream or Grpc over https. There is an open issue about this [here](https://github.com/godotengine/godot/issues/84559) and some people are recommending adding libssl.so manually to the APK. I think this is a major issue which should be fixed before the release as most multiplayer games using C# would be needed a TLS connection. Thank you for all your efforts ahead.

**Comment by van800** [comment]
I looked into debugging C# in the android export once relatively long time ago. https://github.com/godotengine/godot-proposals/issues/3689
Is it worth looking again? Any tips please?

**Comment by raulsntos** [comment]
@van800 Sorry, it's easy to miss comments on closed issues/PRs. I'd love to know more about what issues you have with debugging Android C# applications.

I'm assuming you are asking to support debugging in JetBrains Rider, Godot uses the `linux-bionic` Mono runtime so whatever Rider uses with those kinds of applications should likely work here too.

Feel free to reach out in the [development chat](https://chat.godotengine.org) if you have follow-up questions.

**Comment by ozanyasindogan** [comment]
> @van800 Sorry, it's easy to miss comments on closed issues/PRs. I'd love to know more about what issues you have with debugging Android C# applications.
> 
> I'm assuming you are asking to support debugging in JetBrains Rider, Godot uses the `linux-bionic` Mono runtime so whatever Rider uses with those kinds of applications should likely work here too.
> 
> Feel free to reach out in the [development chat](https://chat.godotengine.org) if you have follow-up questions.

Hi Raul. APK's does not contain libssl dependencies which are required by .Net and any SSL/TLS connection attempt with .Net libraries makes the game crash. The issue explained here #84559, some MS staff even suggested solutions for it and I believe it's a very important bug for .Net developers.
